Feature Proposal

hml-pip opened the pull request:
Regarding the features I [mentioned](https://github.com/ggand0/viewskater/issues/51), I haven't read the entire codebase and unsure if my implementations will negatively affect performance.

- [ ] Compressed File Support
- I'd like to eventually add support for other formats like RAR and 7z. However, I want to ensure the initial ZIP implementation is solid before moving on.
- For large compressed files, we might need a dedicated loading UI. I'm thinking of popup with content similar to [this](https://github.com/iced-rs/iced/tree/master/examples/download_progress).

- [ ]  Fullscreen Window Mode
- I have tested this feature on Windows. I do not have a Linux machine with a GUI, and I use a headless M4 Mac Mini.
- When I connect to the Mac via RustDesk, I don't see a difference between [window.set_simple_fullscreen](https://docs.rs/winit/latest/x86_64-unknown-linux-gnu/winit/platform/macos/trait.WindowExtMacOS.html#tymethod.set_simple_fullscreen) and [window.set_fullscreen](https://docs.rs/winit/latest/x86_64-unknown-linux-gnu/winit/window/struct.Window.html#method.set_fullscreen). I'm unsure if the fullscreen functionality is working as intended on macOS.
- The window title stops updating when switching from fullscreen back to a windowed view. Manually toggling the window's focus (by clicking away and back) fixes this. I've tried to replicate the focus-related code in other places but haven't been successful yet.
- I made some UI tweaks for fullscreen mode in a hackish way, I'm unsure if this is considered good practice within the iced framework.

hml-pip commented:
I've implemented most of the features I need to use this project as my daily image viewer.

My implementation introduces several new variables, which I'm concerned might be considered 'noise.' 
Would you be open to this kind of change?

Compared to Honeyview, I'm not satisfied with the performance I'm getting with compressed files. Are there any simple optimizations or 'low-hanging fruit' that I might be overlooking?

Please let me know if there are any coding style guidelines I should follow or existing problems that need to be addressed.
Thanks again for this great project!

ggand0 commented:
Awesome, I'll look into this!

hml-pip commented:
>     * Linux (Ubuntu 22.04): Zip viewing and fullscreen mode both seem to work well. On Ubuntu, the fullscreen mode hides the titlebar so you can't see the image filename, but this behavior is the same in apps like web browsers and eog image viewer and I think this is totally fine.
The problem I can't fix here is the title won't update after toggle from fullscreen to window, otherwise, I can make it float like the fps bar I [tweaked](https://github.com/ggand0/viewskater/pull/52/commits/757b0bcef176601884ac4e7dccde7c3a06629847) for fullscreen.

 
>     * macOS (M1): For zip viewing, the app crashed as it tried to load metadata files like ones with "._" prefixes. 
What's the full filename for these files?
I can't reproduce this on windows.

>For fullscreen mode, it turns out that winit's `set_simple_fullscreen()` makes the window borderless, while `set_fullscreen()` is the one that makes it fullscreen. No worries about not catching these given your headless setup. I was able to fix both issues with small changes on my macbook pro. I'd like to make commits to your branch in your fork sometime later.

Thank you.

> I noticed the mouse wheel navigation feature and I do like it, but this removes the existing zoom feature. I'd like to revert this change for this PR and do another PR for making it configurable (navigation vs zoom) so user can choose preferred mouse wheel behavior, and we could also support ctrl + "+-" key based zooming as an alternative.
> 
I do make this [work](https://github.com/ggand0/viewskater/pull/52/commits/1b746794b215e61ff426c13beaec977739a6f035) with ctrl/cmd key+mouse wheel, if you prefer the keyboard solution, I can add it too. Just want to make sure the code i write is a good practice within iced.

I was thinking about configuration for users, but this is a complex problem considering cross-platform, As a long time windows user, the problem is already [complicated](https://stackoverflow.com/a/319767). 
If you have any specific need for linux and mac, I can help with using [this](https://docs.rs/crate/config/latest) or other capable crate, I can fix most of the problems on windows.

ggand0 commented:
> The problem I can't fix here is the title won't update after toggle from fullscreen to window

I've fixed this issue by setting window title when the title is actually changed instead of using the `moved` flag
```
let mut last_title = String::new();
let new_title = state.program().title();
if new_title != last_title {
    window.set_title(&new_title);
    last_title = new_title;
}
```

ggand0 commented:
>  the title won't update after toggle from fullscreen to window

I've fixed the issue by setting window title when the title is actually changed instead of using the `moved` flag:
```
let new_title = state.program().title();
if new_title != *last_title {
    window.set_title(&new_title);
    *last_title = new_title;
}
```

hml-pip commented:
Thanks for the fix!

>I excluded files with these prefixes to fix the crash on macOS, but I should also add cfg-gating for macOS.

The root cause of this problem is the method I [implemented](https://github.com/hml-pip/viewskater/blob/00f59dd9f21d29a51a9842874a79829c5d987086/src/cache/cache_utils.rs#L47) for reading image form the compressed file didn't handle all the errors.
I modified your zip file with file `__MACOSX\landscape\._00000000_(2).jpg` renamed to `00000000_(2).jpg` and put it back to `landscape\00000000_(2).jpg` still crash the application, remove it from the zip will fix the problem.

[landscape.zip](https://github.com/user-attachments/files/22001809/landscape.zip)

```rust
pub fn supported_image(name: &str) -> bool {
    // Filter out macOS metadata files
    if name.starts_with("__MACOSX/") {
        return false;
    }
    ALLOWED_EXTENSIONS.contains(&name.split('.').next_back().unwrap_or("").to_lowercase().as_str())
}
```
This should fix your problem for now, the other filter should not be necessary, we need a proper error handler for broken or invalid image in compressed files.
Sorry for this, I don't have a good idea currently.

>Saving and loading a simple YAML file or something would be a good starting point. I'd like to keep the scope of each PR small, so we can discuss this in a separate issue.

My preference for windows is similar with yours, just whatever config file locate in the same folder with executable, it might have permission or other problems. I'll handle the windows side when you're ready to move forward with this.

`VIEWER_MODE`
I made this flag temporary to avoid break your workflow, if you think the viewer behavior is fine, we can remove this flag altogether.

`CTRL_PRESSED` is harder to implement with local variable, I need some time to think about how to implement it properly.

ggand0 commented:
> This should fix your problem for now, the other filter should not be necessary, we need a proper error handler for broken or invalid image in compressed files.

Thanks, I tried viewing the same zip file with the simplified filter condition and it works fine. I added the error handling at image loading functions in [this commit](https://github.com/ggand0/viewskater/pull/52/commits/3dcba9d2f29cd9047de818277aa2b863f7235b30). We could check if each image has valid magic numbers ([ref](https://stackoverflow.com/a/8350345)), but I don't know if they always work for valid images.
  
> My preference for windows is similar with yours, just whatever config file locate in the same folder with executable, it might have permission or other problems. I'll handle the windows side when you're ready to move forward with this.

I use the `dirs` crate to get platform specific user accessible locations to save log files and we can also store config files there. On Windows this will be something like `C:\Users\{username}\AppData\Roaming\viewskater\`. I'll ping you when I have something ready to review.

For the `VIEWER_MODE`: I like the current behavior, but other image viewers like qview and eog on Linux zoom into image just by mouse wheel, so I'd imagine some people might prefer that UI. I'd say keep it but rename it to something more descriptive like `MOUSE_WHEEL_ZOOM` (or mouse_wheel_zoom in the app struct) and we set it to false by default.

> CTRL_PRESSED is harder to implement with local variable, I need some time to think about how to implement it properly.

No rush at all. For example it can be implemented by adding the same fields to ImageShader widget and setting it via a builder method like this:
```rust
impl<Message> ImageShader<Message> {
  ...
  pub fn with_interaction_state(mut self, mouse_wheel_zoom: bool, ctrl_pressed: bool) -> Self {
      self.mouse_wheel_zoom = mouse_wheel_zoom;
      self.ctrl_pressed = ctrl_pressed;
      self
  }
}
```
Then we can call it in the ui module for single pane mode
```rust
let shader = ImageShader::new(Some(scene))
  .with_interaction_state(app.panes[0].mouse_wheel_zoom, app.panes[0].ctrl_pressed)
  .width(Length::Fill)
  .height(Length::Fill)
  .content_fit(iced_winit::core::ContentFit::Contain)
  .horizontal_split(false);
```
For the dual-pane mode, we'd need to propagate the states to Pane objects but the same principle. Would you like to work on this?

ggand0 commented:
The ctrl/cmd + mouse wheel zooming and the menu behavior toggle work well on macOS and Linux with your latest commit.

> Could you make the ImageCache removable?

I can make it removable. For invalid images, I'd like to explore what's the best practice for handling them on a separate branch, as this is a general issue affecting both regular directories and compressed files.

I looked into the performance issue you mentioned earlier:
> Compared to Honeyview, I'm not satisfied with the performance I'm getting with compressed files. Are there any simple optimizations or 'low-hanging fruit' that I might be overlooking?

I tested the performance with compressed files (zip, rar, 7z solid, 7z non-solid) on Linux (uploaded the files on [GDrive](https://drive.google.com/drive/u/0/folders/1o9ZJZZQvEKJSx7O_M2P2Nivbzyh1HL8A)), and other than 7z solid it was indeed slower. I created the 7z files like this:
```bash
# solid (files compressed together as one stream, better compression)
7z a small_images_linux_solid.7z ./small_images/

# non-solid (files compressed individually, faster random access)
7z a -ms=off small_images_linux_nonsolid.7z ./small_images/
```
																												   
I noticed that `read_compressed_bytes_by_name` opens the full archive file from disk for every image load, since the global `COMPRESSED_FILES` only stores the path and type. I tried caching the archive instance itself in the Pane struct: `pub archive_cache: Arc<Mutex<ArchiveCache>>,` with some refactoring, and it's much faster now. This way we can also avoid the global variable `COMPRESSED_FILES`. 

For RAR files, the while loop-based file search seems to be necessary due to the limitation of this format.
With the `small_images_linux.rar` file, keyboard-based rendering performance was around 17 FPS in skate mode. After implementing archive instance caching, it improved to 20-22 FPS, but it was still slow. Then I realized that there was this `debug!("reading rar {} ?= {:?}", filename, entry_filename);` logging line inside the while loop. After removing this line, it now gets ~60 FPS.

Would you mind if I make a commit for this?

ggand0 commented:
Thanks, I've implemented per-pane archive instance caching and made some performance improvements. Here are the FPS values I'm seeing with the "small_images" dataset in `opt-dev` profile when skating through the first 500 images on Ubuntu with a 7900XTX GPU:
- regular directory: 180-200 fps
- zip: 180-200 fps
- rar: 60+ fps
- non-solid 7z: 110-120 fps
- solid 7z:  180-200 fps

Compressed file viewing now works with the dual-pane mode too. For solid 7z files this loads the entire archive file data into memory, so I'll add logic to fall back to the original behavior (reopening archive file on every image load) when it's too large. 

There are two remaining issues:
1. The slider-based rendering doesn't work with solid 7z files on Linux (laggy and nothing is rendered during the slider movement)
2. The app crashes when trying to load larger solid 7z files on macOS, but this doesn't happen on Linux. When I try to open a solid 7z file created from the `small_images` dir on macOS, I get this crash:
<details>
<summary>macOS crash log details</summary>

```
2025-08-31T10:42:49.336614Z DEBUG viewskater::app:1302 File dropped: ["/Users/gota/ggando/vs_data/small_images_macos_solid.7z"]
2025-08-31T10:42:49.336664Z DEBUG viewskater::app:241 last_opened_pane: 0
2025-08-31T10:42:49.336671Z DEBUG viewskater::app:268 pane_file_lengths: [0]
2025-08-31T10:42:49.336677Z INFO viewskater::utils::mem:101 MEMORY [60.66MB] - Before pane initialization
2025-08-31T10:42:49.400821Z DEBUG viewskater::pane:711 "/Users/gota/ggando/vs_data/small_images_macos_solid.7z" block_count: 1
2025-08-31T10:42:49.400858Z DEBUG viewskater::pane:716 Using 8 threads to read "/Users/gota/ggando/vs_data/small_images_macos_solid.7z"


[PANIC] at src/pane.rs:736 - panicked at src/pane.rs:736:22:
Failed block reading 7z file: Io(Custom { kind: Other, error: ChecksumVerificationFailed }, "")
[PANIC] Backtrace:
[BACKTRACE] 0: backtrace::backtrace::trace
[BACKTRACE] 1: backtrace::capture::Backtrace::new
[BACKTRACE] 2: viewskater::file_io::setup_panic_hook::{{closure}}
[BACKTRACE] 3: std::panicking::rust_panic_with_hook
[BACKTRACE] 4: std::panicking::begin_panic_handler::{{closure}}
[BACKTRACE] 5: std::sys::backtrace::__rust_end_short_backtrace
[BACKTRACE] 6: __rustc::rust_begin_unwind
[BACKTRACE] 7: core::panicking::panic_fmt
[BACKTRACE] 8: core::result::unwrap_failed
[BACKTRACE] 9: std::sys::backtrace::__rust_begin_short_backtrace
[BACKTRACE] 10: core::ops::function::FnOnce::call_once{{vtable.shim}}
[BACKTRACE] 11: std::sys::pal::unix::thread::Thread::new::thread_start
[BACKTRACE] 12: __pthread_cond_wait

A complete crash log has been written to: /Users/gota/Library/Application Support/viewskater/logs/panic.log
```
</details>


This crash didn't happen when I tried another solid 7z file that has the first 1,000 images from the `small_images` data. I haven't looked into the details of this yet, but it might be related to [this issue](https://github.com/hasenbanck/sevenz-rust2/issues/46) of the sevenz-rust2 crate.
	
Can you try running this updated code on Windows and see if it works?

ggand0 commented:
I've fixed the slider navigation lag for solid 7z files. It was cloning the entire list of `PathTyle::FileByte` elements.
	
I noticed the `PathType::FileByte enum` is only used for solid 7z. We can simplify this by reverting to `image_paths: Vec<PathBuf>` in ImageCache and moving the solid 7z pre-loaded data as a HashMap. `file_io::read_image_bytes()` can be moved there as a method too.

> For the performance of 7z, It seems the file [created](https://mega.nz/folder/oLYwjCzD#QqkLFYcL5cp_sT6m996jcQ) with different codec will have mixed results, the solid file created without using LZMA2 will be terrible when reading the entire file.

This is a good find. LZMA2 seems to be the default codec for 7z, so it's fine if it runs with that codec. When I tested on Linux, the bzip2_solid and non_solid files ran without lag though.

Nice, your latest solid 7z loading snippet works on macOS, and now I can open the solid 7z file that was causing the crash before.
I feel like the PR is ready for merge once we have added the fallback logic based on file size limits for solid 7z and completed the refactoring I mentioned.

ggand0 commented:
> I'm not sure loading files on the fly is a good solution for zip/rar under all circumstances.

We can keep the current code that stores all the image paths for zip/rar files. I was talking about the solid 7z format. 

I tested the performance without preload for solid 7z files, and it was super slow and unusable. So we do want to preload them as much as we can. I want to address cases where the solid 7z (and non-solid 7z if you plan to store file data in memory for some specific codecs) file size is too large to fit into the memory like more than a few GB. Having the OS crash or slow down significantly when user opens a large solid 7z file in the app (due to the memory pressure) is not ideal. Perhaps we could
1. show a warning dialog when the file size of solid 7z exceeds a threshold (e.g., 1GB)
or
2. stop loading when the loaded data exceeds the threshold in the preload process.

This threshold should be configurable by users once we have implemented the config feature too. I'm leaning toward the option 2 but let me know what you think.

I'll work on this refactoring soon:
> I noticed the `PathType::FileByte enum` is only used for solid 7z. We can simplify this by reverting to `image_paths: Vec<PathBuf>` in ImageCache and moving the solid 7z pre-loaded data to the ArchiveCache struct as a HashMap. `file_io::read_image_bytes()` can be moved there as a method too.

hml-pip commented:
Sorry for the RAR bug, I should have done a proper test before pushing it.

>I've added the warning dialog for solid 7z using rfd::MessageDialog which should be cross-platform. I confirmed it works on Linux and macOS.

Works on Windows.

>The compressed file viewing feature works now for most cases, and I'd like to merge this PR soon. I get a "ViewSkater not responding" dialog on Linux when loading a 600MB+ solid 7z file on the initial load, but it's functional and we do warn users when opening large solid 7z files, so I think we should address this in a separate PR.

The results is varied with different compressed files, this will happen on most 7z files not created with  `LZMA2` larger than certain size. If the file is slow to decompress(solid or non `LZMA2/BZIP2`) it affect the slider performance if you don't preload it.
I developed [this](https://github.com/ggand0/viewskater/pull/52/commits/38044a8c26fdcdd3279354285af874569d8557a2) to work around the slider problem, for ZIP/RAR files, it only seems to benefit certain scenarios.
I'll make another PR if I want to address this since it needs a lot of test data for future reference.

>Are you aware of any critical issues that should be addressed in this PR

1. `Copy file path` button generate `Z:\testdata\bzip2_non_solid.7z/00000870.jpg` on Windows.
I'm not sure this should fix or not, this is not a useful feature for my use case.
2. I'd like to add compressed file name to title for future reference, user might find it useful?

I'll report later when I have more test data available.

>This thread has gotten quite long, so I'd like to avoid adding any more features like minor performance improvements and focus on completing it.

Could you give me a general idea of how small future PRs should be?"

hml-pip commented:
Sorry for the RAR bug, I should have done a proper test before pushing it.

>I've added the warning dialog for solid 7z using rfd::MessageDialog which should be cross-platform. I confirmed it works on Linux and macOS.

Works on Windows.

>The compressed file viewing feature works now for most cases, and I'd like to merge this PR soon. I get a "ViewSkater not responding" dialog on Linux when loading a 600MB+ solid 7z file on the initial load, but it's functional and we do warn users when opening large solid 7z files, so I think we should address this in a separate PR.

The results are varied with different compressed files, this will happen on most 7z files not created with  `LZMA2` larger than certain size. Solid or not, it  affect the slider or initial loading performance.
I developed [this](https://github.com/ggand0/viewskater/pull/52/commits/38044a8c26fdcdd3279354285af874569d8557a2) to work around the slider problem, for ZIP/RAR files, it only seems to benefit certain scenarios.
I'll make another PR if I want to address this since it needs a lot of test data for future reference.

>Are you aware of any critical issues that should be addressed in this PR

1. `Copy file path` button generate `Z:\testdata\bzip2_non_solid.7z/00000870.jpg` on Windows.
I'm not sure this should fix or not, this is not a useful feature for my use case.
2. I'd like to add compressed file name to title for future reference, user might find it useful?

I'll report later when I have more test data available.

>This thread has gotten quite long, so I'd like to avoid adding any more features like minor performance improvements and focus on completing it.

Could you give me a general idea of how small future PRs should be?"

hml-pip commented:
```rust
fn read_7z_path(path: &PathBuf, file_paths: &mut Vec<PathSource>, archive_cache: &mut ArchiveCache) -> Result<(), Box<dyn Error>> {
    use std::thread;
    use std::io::Read;
    let password = sevenz_rust2::Password::empty();
    let mut file = File::open(path)?;
    let archive = sevenz_rust2::Archive::read(&mut file, &password)?;
    let is_solid = archive.is_solid;
    let mut files = Vec::new();
    let mut image_names = Vec::new();
    // Set up the archive cache for this 7Z file
    archive_cache.set_current_archive(path.clone(), ArchiveType::SevenZ);

    // First pass: collect all image files and their sizes
    for entry in archive.files.iter() {
        if !entry.is_directory && supported_image(entry.name()) {
            files.push(entry.size());
            image_names.push(entry.name());
        }
    }

    let image_size = files.iter().sum::<u64>();
    debug!("Total image size: {}mb", image_size / 1_000_000);
    // Determine if we'll preload this archive (small archives get preloaded)
    let will_preload = is_solid || image_size < CONFIG.archive_cache_size;

    // Check for large solid archives and show warning dialog
    if will_preload && image_size > 0 {
        let archive_size_mb = image_size / 1_000_000;

        // Show warning dialog for archives larger than configured threshold
        if archive_size_mb > CONFIG.archive_warning_threshold_mb {
            let (available_gb, is_recommended) = mem::check_memory_for_archive(archive_size_mb);

            // Show the warning dialog and check user response
            if !file_io::show_memory_warning_sync(archive_size_mb, available_gb, is_recommended) {
                // User chose not to proceed
                return Err("User cancelled loading large solid 7z archive".into());
            }
            // User chose to proceed, continue with loading
        }

        // solid file is too slow for lazy loading - proceed with preload
        let block_count = archive.blocks.len();
        debug!("{path:?} block_count: {block_count}");
        let cpu_threads = if thread::available_parallelism().is_ok() {
            thread::available_parallelism()?.get() as u32
        } else { 4 };

        debug!("Using {cpu_threads} threads to read {path:?}");
        let sevenz_data = Mutex::new(Vec::new());
        for block_index in 0..block_count {
            thread::scope(|s| {
                s.spawn(||{
                    let mut source = File::open(path).unwrap();
                    // 2. For decoders that supports it, we can set the thread_count on the block decoder
                    //    so that it uses multiple threads to decode the block. Currently only LZMA2 is
                    //    supporting this. We try to use all threads report from std::thread.
                    let block_decoder = sevenz_rust2::BlockDecoder::new(cpu_threads, block_index, &archive, &password, &mut source);
                    block_decoder.for_each_entries(&mut |entry, reader| {
                        let mut buffer = Vec::new();
                        if !entry.is_directory && supported_image(entry.name()) {
                            reader.read_to_end(&mut buffer)?;
                            sevenz_data.lock().unwrap().push((entry.name().to_string(), buffer));
                        } else {
                            // As `for_each_entries` noted, we can not skip any files we don't want.
                            // Discard all the bytes we don't need.
                            let _ = std::io::copy(&mut reader.take(entry.size()), &mut std::io::sink());
                        }
                        Ok(true)
                    })
                    .expect("Failed block reading 7z file");
                });
            });
        }

        // Add files and preloaded data to respective structures
        let data_list = sevenz_data.into_inner()?;
        for (filename, data) in data_list {
            let path_buf = PathBuf::from(&filename);
            // Solid 7z archives are always preloaded - use Preloaded variant
            file_paths.push(PathSource::Preloaded(path_buf));
            archive_cache.add_preloaded_data(filename, data);
        }
    } else {
        // Non-perload 7z: just list files without preloading - use Archive variant
        for name in &image_names {
            file_paths.push(PathSource::Archive(PathBuf::from(name)));
        }
    }

    Ok(())
}
```
This use the same method with RAR/ZIP for caching, ignore 7z files that contain no images.
Change the `read_to_end` to [this](https://stackoverflow.com/questions/42243355/how-to-advance-through-data-from-the-stdioread-trait-when-seek-isnt-impleme), but I don't see any performance improvement.

ggand0 commented:
> 3. [This](https://github.com/ggand0/viewskater/blob/8750b3e3e526c853c014d07c12ff8e86b54af135/src/pane.rs#L430) break the file sorting and remove the directory name from file in the compressed file.

Thanks for pointing it out. Your PathSource::file_name fixes the issue on my machine too. I have tested the new read_7z_path() and it works. 1 is fine and for 2 you could add it in the next PR (tiny PRs are welcome too). Can you confirm that the compressed file viewing feature works with zip/rar/small 7z files (ones compressed with LZMA2) on your machine now?

> Could you give me a general idea of how small future PRs should be?

This is something I'm still learning too, and I also tend to create longer PRs when I work on things myself. But I realized it becomes a bigger issue when you collaborate with others. I think you can ship more features with smaller PRs and the review process will be faster and more accurate ([ref](https://graphite.dev/blog/code-review-best-practices)). For example, I'd split this PR into three:
**PR 1**: Fullscreen mode feature
**PR 2**: Basic compressed file support (ZIP + RAR, RAR performance issue was easy to fix)
**PR 3**: 7z support + optimizations (harder to fix the slider performance bug + remaining issues for non LZMA2 codecs)
